### ISSUE TITLE
Incompatible equality test of addresses in Distributed 2021.3.0

### ISSUE THREAD

**Issue by rainwoodman** [dask/distributed#4565]
**What happened**:

After updating from 2021.2.0 to 2021.3.0, our regular tests started to observe this error. 
Any pointers as of what might be a good way to mitigate the issue?

````
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/yfeng1/source/nbodykit/build/testenv/lib/python3.8/site-packages/nbodykit/tests/test_distributed.py:32: in test_save
    cat.save(tmpfile, dataset='data')
/home/yfeng1/source/nbodykit/build/testenv/lib/python3.8/site-packages/nbodykit/base/catalog.py:692: in save
    source.store(target, regions=region, lock=False, compute=True)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/array/core.py:1540: in store
    r = store([self], [target], **kwargs)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/array/core.py:1041: in store
    result.compute(**kwargs)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:283: in compute
    (result,) = compute(self, traverse=False, **kwargs)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:553: in compute
    schedule = get_scheduler(
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:1151: in get_scheduler
    return get_scheduler(scheduler=config.get("scheduler", None))
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:1141: in get_scheduler
    return get_client().get
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/distributed/worker.py:3200: in get_client
    return worker._get_client(timeout=timeout)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/distributed/worker.py:3076: in _get_client
    or client._start_arg == self.scheduler.address
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = LocalCluster(d66622f8, 'inproc://192.168.86.31/163961/9', workers=1, threads=1, memory=16.59 GB)
other = 'inproc://192.168.86.31/163961/1'

    def __eq__(self, other):
>       return self.name == other.name
E       AttributeError: 'str' object has no attribute 'name'

/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/distributed/deploy/cluster.py:461: AttributeError
```

**What you expected to happen**:

It is not clear how we could triggered malformed input. We are simply creating a LocalCluster in this case. 
Might be a regression on distributed's side. 

**Minimal Complete Verifiable Example**:

```python
# Put your MCVE code here
```

**Anything else we need to know?**:

**Environment**:

- Dask version:  2021.3.0
- Python version: 3.8
- Operating System: Linux
- Install method (conda, pip, source):
Conda

**Comment by jrbourbeau** [comment]
Thanks for raising an issue @rainwoodman. It's not immediately clear to me what in the latest release could be causing this. The particular `AttributeError` which is being triggered is happening in some code which was added in 2021.01.1. Do you have a code snippet you could post that I could run?

**Comment by rainwoodman** [comment]
Thanks for looking into it.

I actually have had difficulty reproducing this on my local machine. It used to occur randomly on our github workflow, but has become a persistent error recently:

https://pipelines.actions.githubusercontent.com/4WG3RDC5LrvSQSfVuzaeJ6lLFmXDn3iE4xIDQGf4ZMGoiw0ciN/_apis/pipelines/1/runs/206/signedlogcontent/3?urlExpires=2021-05-08T09%3A25%3A08.0077385Z&urlSigningMethod=HMACV1&urlSignature=NWLiHD2e9K0Xp9e6ZAskBUQpF5SX9bi5rds55o4zYKU%3D

I'll try to contain this error more closely and report back.

**Comment by fjetter** [comment]
I could reproduce and probably fix the issue in #4799. It is an application I would not consider straight forward, though :) Either way, I got the edge case and wrote a test for it. No idea whether this has been connected to any recent change, though.
